feat: Add deal parameters (#778)
eigerco/polka-storage: docs/src/architecture/pallets/market.md

## Table of Contents

- [Market Pallet](#market-pallet)

>   - [Table of Contents](#table-of-contents)

Reviewer: These two are redundant

Author: Auto formatting :angry:

Reviewer: I know, I used to use the same plugin :P

[^settle_deal_payments]: Read more about the `settle-deal-payments` command in [_Storagext CLI/Subcommand `market`/`settle-deal-payments`_](../../storagext-cli/market.md#settle-deal-payments)

### `publish_deal_parameters`

This extrinsic lets the storage provider publish or update deal parameters.

| Name                      | Description                                                                       | Type                                                     |
| ------------------------- | --------------------------------------------------------------------------------- | -------------------------------------------------------- |
| `minimum_price_per_block` | The minimum price per block the SP is willing to accept                           | Positive integer, in [Plancks](../../glossary.md#planck) |

> | `deal_duration`           | The optional lower and upper bound of a deal duration the SP is willing to accept | `DealDurationBound`                                      |

Reviewer: In tandem with the comment below, I believe this makes more sense
```suggestion
| `deal_duration`           | The lower and upper bounds of a deal duration the SP is willing to accept | `DealDurationBound`                                      |
```

Where `DealDurationBound` consists of:

| Name  | Description                                        | Type             |
| ----- | -------------------------------------------------- | ---------------- |
| lower | Shortest deal duration the SP is willing to accept | Positive integer |

> | upper | Longest deal duration the SP is willing to accept  | Positive integer |

Reviewer: ```suggestion
| `lower` | Shortest deal duration the SP is willing to accept | Positive integer |
| `upper` | Longest deal duration the SP is willing to accept  | Positive integer |
```

> | `upper` | Longest deal duration the SP is willing to accept  | Positive integer |

Reviewer: Maybe we should note that they're optional.

### `remove_deal_parameters`

A signed extrinsic allows a storage provider to remove the deal parameter they have previously set.

> If no deal parameters are present this extrinsic will not fail.

Reviewer: This *might* be more accurate (you're the judge)
```suggestion
If no deal parameters are present this extrinsic is a no-op.
```

Author: [It is only a no-op if it returns an error](https://crates.parity.io/frame_support/macro.assert_noop.html). I did consider making this fail explicitly if there are no parameters present, might save on gas costs if SP double removes on accident.

Reviewer: Then, if there are no parameters, it replaces the existing ones with "empty ones"?

Author: No it doesn't replace, I attempts to [remove](https://docs.rs/frame-support/latest/frame_support/storage/types/struct.StorageMap.html#method.remove), which is infallible so I am assuming if nothing is present it does nothing, I didn't look into the implementation.

| ------------------ | ----------------------------------------------------------------------------- | ------------------------- |
| `MaxDeals`         | How many deals can be published in a single batch of `publish_storage_deals`. | 128                       |
| `MaxDealsPerBlock` | Maximum deals that can be scheduled to start at the same block.               | 128                       |

> | `MinDealDuration`  | Minimum time an activated deal should last.                                   | 5 Minutes (50 Blocks)     |

Reviewer: Any chance we could get a md formatting tool to run so we stop having these?

Author: [markdownlint](https://www.npmjs.com/package/markdownlint) is an option for this. My VSCode uses this for liting too
